qiime2/sdk/result.py

```diff
@@ -282,6 +286,205 @@ def result(self):
+        self._annotations = []
+        annotations_dir = self._archiver.annotations_dir
+        # Not sure if there's a better way to do this, but this accounts for
+        # the fact that the annotations_dir will be None for all
+        # previous archive versions < 7.0
+        if annotations_dir and os.path.exists(annotations_dir):
+            for annotation in os.listdir(annotations_dir):
+                self._annotations.append(
+                    Annotation.load(os.path.join(annotations_dir, annotation))
+                )
+
+    def add_annotation(self, annotation):
+        """Add an Annotation onto a Result object.
+        All Result-associated parameters are passed into the sub-class's
+        `write` method, while the Annotation instance handles everything else.
+
+        Parameters
+        ----------
+        annotation\n
+            An instantiated Annotation subclass (Note, etc).
+
+        Raises
+        ------
+        ValueError\n
+            If the Annotation name matches an existing Annotation name
+            attached to the Result in question.
+
+        Notes
+        -----
+            In Archive Format 7.0, `referenced_result_uuid` is set to
+            the same value as `root_result_uuid`, but this will change
+            in future versions to allow for Annotations that may reference
+            a different Result than the one they are attached to.
+
+        See Also
+        --------
+        Annotation.write
+
+        """
+        self._validate_annotation_support()
+        # Guard to ensure Annotation names are unique per Result object
+        for existing_annotation in self._annotations:
+            if annotation.name == existing_annotation.name:
+                raise ValueError(
+                    'Namespace collision occurred when attempting to add '
+                    f'Annotation with name: "{annotation.name}"\n'
+                    'Annotation names must be unique within each Result '
+                    'they are attached to.'
+                )
+
+        annotation.write(annotations_dir=self._archiver.annotations_dir,
+                         root_result_uuid=str(self.uuid),
+                         referenced_result_uuid=str(self.uuid))
+        self._annotations.append(annotation)
+
+        # now calculate checksums for all files within the newly minted
+        # annotation subdir
+        annotation_dir = \
+            pathlib.Path(self._archiver.annotations_dir) / str(annotation.uuid)
+        checksum_ext = self._archiver._fmt.CHECKSUM_TYPE
+        manifest = self._archiver._fmt.CHECKSUM_FILE
+
+        checksums = util.checksum_directory(str(annotation_dir),
+                                            checksum_type=checksum_ext)
+        checksums.pop(manifest, None)
+
+        with (annotation_dir / manifest).open('w') as fh:
+            for item in checksums.items():
+                fh.write(util.to_checksum_format(*item))
+                fh.write('\n')
```

Comment:
It seems like this could all go in `Annotation.write` was it so that you can use `CHECKSUM_TYPE`? You could probably make those arguments to write (presuming it is converted to private) or just hardcode them to sha512.

Reply:
this seemed like a reasonable place to add this since the checksum logic happens separately from the actual writing of the annotation - would you prefer it lives under `_write` though? if that feels more sensible i can move it there.

Reply:
evan and i discussed offline and are going to leave this here until a future time dictates this being more sensible to move elsewhere... bc decision fatigue lol